ext/standard/var.c

```diff
@@ -1209,6 +1235,16 @@ static void php_var_serialize_intern(smart_str *buf, zval *struc, php_serialize_
 				 && Z_OBJ_HT_P(struc)->get_properties == zend_std_get_properties) {
 					/* Optimized version without rebulding properties HashTable */
 					zend_object *obj = Z_OBJ_P(struc);
+
+					if (zend_lazy_object_must_init(Z_OBJ_P(struc))
+							&& zend_lazy_object_initialize_on_serialize(Z_OBJ_P(struc))) {
+						obj = zend_lazy_object_init(Z_OBJ_P(struc));
+						if (!obj) {
+							smart_str_appendl(buf, "N;", 2);
```

Comment:
Good point. This was an attempt to copy what serialize() does in other cases of failures (e.g. after calling __sleep), but I misunderstood it (it only adds `N;` on failure when there is no exception). We can leave out the `N;` indeed.

> Optimally we'd immediately abort serialization, but that might require more changes and checks that slow down the happy path.
> That said, not handling exceptions also means that we might trigger serialize() calls even when there was an exception.

Yes. I think that serialize() achieves this by returning early in `php_var_serialize_intern()` when `EG(exception)` is set. There are some cases where it will continue serialization to some degree (e.g. when serializing an array), but it will not recurse or attempt to serialize any object.